File: integration_tests/java_client_tests/client-app/src/test/java/dev/graphlink/test/SubscriptionsTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkException;
import dev.graphlink.test.generated.client.GraphLinkSubscriptionListener;
import dev.graphlink.test.generated.enums.UserStatus;
import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.types.UserCreatedResponse;
import dev.graphlink.test.generated.types.UserStatusChangedResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class SubscriptionsTest {

    private MockAdapter adapter;
    private MockWebSocketAdapter wsAdapter;
    private GraphLinkClient client;

    @BeforeEach
    void setUp() {
        adapter = new MockAdapter();
        wsAdapter = new MockWebSocketAdapter();
        client = newClient(adapter, wsAdapter);
    }

    // ── userCreated ────────────────────────────────────────────────────────────

    @Nested
    class UserCreated {

        @Test
        void singleEvent_deserializedCorrectly() {
            List<UserCreatedResponse> events = new ArrayList<>();
            client.subscriptions.userCreated(onMessage(events));

            wsAdapter.pushEvent(Map.of("userCreated", kUserAliceJson));

            assertEquals(1, events.size());
            assertEquals("user-1", events.get(0).getUserCreated().getId());
        }

        @Test
        void singleEvent_name_isCorrect() {
            List<UserCreatedResponse> events = new ArrayList<>();
            client.subscriptions.userCreated(onMessage(events));

            wsAdapter.pushEvent(Map.of("userCreated", kUserAliceJson));

            assertEquals("Alice Smith", events.get(0).getUserCreated().getName());
        }

        @Test
        void singleEvent_status_deserializesToEnum() {
            List<UserCreatedResponse> events = new ArrayList<>();
            client.subscriptions.userCreated(onMessage(events));

            wsAdapter.pushEvent(Map.of("userCreated", kUserAliceJson));

            assertEquals(UserStatus.ACTIVE, events.get(0).getUserCreated().getStatus());
        }

        @Test
        void multipleEvents_deliveredInOrder() {
            List<UserCreatedResponse> events = new ArrayList<>();
            client.subscriptions.userCreated(onMessage(events));

            wsAdapter.pushEvent(Map.of("userCreated", kUserAliceJson));
            wsAdapter.pushEvent(Map.of("userCreated", kUserBobJson));

            assertEquals(2, events.size());
            assertEquals("user-1", events.get(0).getUserCreated().getId());
            assertEquals("user-2", events.get(1).getUserCreated().getId());
        }

        @Test
        void completeMessage_callsOnComplete() {
            AtomicBoolean completed = new AtomicBoolean(false);
            client.subscriptions.userCreated(new GraphLinkSubscriptionListener<UserCreatedResponse>() {
                @Override public void onMessage(UserCreatedResponse r) {}
                @Override public void onComplete() { completed.set(true); }
            });

            wsAdapter.pushComplete();

            assertTrue(completed.get());
        }

        @Test
        void completeMessage_afterEvent_callsOnComplete() {
            List<UserCreatedResponse> events = new ArrayList<>();
            AtomicBoolean completed = new AtomicBoolean(false);
            client.subscriptions.userCreated(new GraphLinkSubscriptionListener<UserCreatedResponse>() {
                @Override public void onMessage(UserCreatedResponse r) { events.add(r); }
                @Override public void onComplete() { completed.set(true); }
            });

            wsAdapter.pushEvent(Map.of("userCreated", kUserAliceJson));
            wsAdapter.pushComplete();

            assertEquals(1, events.size());
            assertTrue(completed.get());
        }

        @Test
        void subscribeMessage_isSent_afterHandshake() {
            client.subscriptions.userCreated(onMessage(new ArrayList<>()));

            long subscribeCount = wsAdapter.getSentMessages().stream()
                    .filter(m -> "subscribe".equals(m.get("type")))
                    .count();
            assertEquals(1, subscribeCount);
        }

        @Test
        void connectionInit_isSent_firstMessage() {
            client.subscriptions.userCreated(onMessage(new ArrayList<>()));

            assertEquals("connection_init", wsAdapter.getSentMessages().get(0).get("type"));
        }
    }

    // ── userStatusChanged ──────────────────────────────────────────────────────

    @Nested
    class UserStatusChanged {

        @Test
        void event_deserializedCorrectly() {
            List<UserStatusChangedResponse> events = new ArrayList<>();
            client.subscriptions.userStatusChanged("user-1", onMessage(events));

            wsAdapter.pushEvent(Map.of("userStatusChanged", kUserAliceJson));

            assertEquals(1, events.size());
            assertEquals("user-1", events.get(0).getUserStatusChanged().getId());
        }

        @Test
        void event_status_isCorrect() {
            List<UserStatusChangedResponse> events = new ArrayList<>();
            client.subscriptions.userStatusChanged("user-1", onMessage(events));

            wsAdapter.pushEvent(Map.of("userStatusChanged", kUserAliceJson));

            assertEquals(UserStatus.ACTIVE, events.get(0).getUserStatusChanged().getStatus());
        }

        @Test
        void userId_sentInSubscribePayload() {
            client.subscriptions.userStatusChanged("user-42", onMessage(new ArrayList<>()));

            Map<String, Object> subscribeMsg = wsAdapter.getSentMessages().stream()
                    .filter(m -> "subscribe".equals(m.get("type")))
                    .findFirst()
                    .orElseThrow();

            @SuppressWarnings("unchecked")
            Map<String, Object> payload = (Map<String, Object>) subscribeMsg.get("payload");
            @SuppressWarnings("unchecked")
            Map<String, Object> variables = (Map<String, Object>) payload.get("variables");
            assertEquals("user-42", variables.get("userId"));
        }

        @Test
        void completeMessage_callsOnComplete() {
            AtomicBoolean completed = new AtomicBoolean(false);
            client.subscriptions.userStatusChanged("user-1", new GraphLinkSubscriptionListener<UserStatusChangedResponse>() {
                @Override public void onMessage(UserStatusChangedResponse r) {}
                @Override public void onComplete() { completed.set(true); }
            });

            wsAdapter.pushComplete();

            assertTrue(completed.get());
        }
    }

    // ── error handling ─────────────────────────────────────────────────────────

    @Nested
    class ErrorHandling {

        @Test
        void errorMessage_callsOnError() {
            AtomicReference<GraphLinkException> caught = new AtomicReference<>();
            client.subscriptions.userCreated(new GraphLinkSubscriptionListener<UserCreatedResponse>() {
                @Override public void onMessage(UserCreatedResponse r) {}
                @Override public void onError(GraphLinkException error) { caught.set(error); }
            });

            wsAdapter.pushError(wsAdapter.getSubscriptionId(),
                    List.of(Map.of("message", "Unauthorized")));

            assertNotNull(caught.get());
        }
    }

    // ── helpers ────────────────────────────────────────────────────────────────

    private static <T> GraphLinkSubscriptionListener<T> onMessage(List<T> collector) {
        return new GraphLinkSubscriptionListener<T>() {
            @Override public void onMessage(T response) { collector.add(response); }
        };
    }
}

File: integration_tests/kotlin_client_tests/client-app/src/test/kotlin/dev/graphlink/test/SubscriptionsTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.enums.UserStatus
import dev.graphlink.test.generated.types.UserCreatedResponse
import dev.graphlink.test.generated.types.UserStatusChangedResponse
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.kUserAliceJson
import dev.graphlink.test.Fixtures.kUserBobJson
import dev.graphlink.test.Fixtures.newClient

class SubscriptionsTest {

    private lateinit var adapter: MockAdapter
    private lateinit var wsAdapter: MockWebSocketAdapter
    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        adapter = MockAdapter()
        wsAdapter = MockWebSocketAdapter()
        client = newClient(adapter, wsAdapter)
    }

    // ── userCreated ────────────────────────────────────────────────────────────

    @Nested
    inner class UserCreated {

        @Test
        fun singleEvent_deserializedCorrectly() = runTest {
            val events = mutableListOf<UserCreatedResponse>()
            val job = launch { client.subscriptions.userCreated().collect { events += it } }
            advanceUntilIdle()

            wsAdapter.deliverConnectionAck()
            val subId = wsAdapter.getLastSubscriptionId()
            wsAdapter.deliverNext(subId, mapOf("userCreated" to kUserAliceJson))
            advanceUntilIdle()

            assertEquals(1, events.size)
            assertEquals("user-1", events[0].userCreated.id)
            job.cancel()
        }

        @Test
        fun singleEvent_name_isCorrect() = runTest {
            val events = mutableListOf<UserCreatedResponse>()
            val job = launch { client.subscriptions.userCreated().collect { events += it } }
            advanceUntilIdle()

            wsAdapter.deliverConnectionAck()
            val subId = wsAdapter.getLastSubscriptionId()
            wsAdapter.deliverNext(subId, mapOf("userCreated" to kUserAliceJson))
            advanceUntilIdle()

            assertEquals("Alice Smith", events[0].userCreated.name)
            job.cancel()
        }

        @Test
        fun singleEvent_status_deserializesToEnum() = runTest {
            val events = mutableListOf<UserCreatedResponse>()
            val job = launch { client.subscriptions.userCreated().collect { events += it } }
            advanceUntilIdle()

            wsAdapter.deliverConnectionAck()
            val subId = wsAdapter.getLastSubscriptionId()
            wsAdapter.deliverNext(subId, mapOf("userCreated" to kUserAliceJson))
            advanceUntilIdle()

            assertEquals(UserStatus.ACTIVE, events[0].userCreated.status)
            job.cancel()
        }

        @Test
        fun multipleEvents_deliveredInOrder() = runTest {
            val events = mutableListOf<UserCreatedResponse>()
            val job = launch { client.subscriptions.userCreated().collect { events += it } }
            advanceUntilIdle()

            wsAdapter.deliverConnectionAck()
            val subId = wsAdapter.getLastSubscriptionId()
            wsAdapter.deliverNext(subId, mapOf("userCreated" to kUserAliceJson))
            wsAdapter.deliverNext(subId, mapOf("userCreated" to kUserBobJson))
            advanceUntilIdle()

            assertEquals(2, events.size)
            assertEquals("user-1", events[0].userCreated.id)
            assertEquals("user-2", events[1].userCreated.id)
            job.cancel()
        }

        @Test
        fun connectionInit_isSent_firstMessage() = runTest {
            val job = launch { client.subscriptions.userCreated().collect {} }
            advanceUntilIdle()

            val first = wsAdapter.sentMessagesDecoded.firstOrNull()
            assertNotNull(first)
            assertEquals("connection_init", first["type"])
            job.cancel()
        }

        @Test
        fun subscribeMessage_isSent_afterConnectionAck() = runTest {
            val job = launch { client.subscriptions.userCreated().collect {} }
            advanceUntilIdle()

            wsAdapter.deliverConnectionAck()

            val subscribeCount = wsAdapter.sentMessagesDecoded.count { it["type"] == "subscribe" }
            assertEquals(1, subscribeCount)
            job.cancel()
        }
    }

    // ── userStatusChanged ──────────────────────────────────────────────────────

    @Nested
    inner class UserStatusChanged {

        @Test
        fun event_deserializedCorrectly() = runTest {
            val events = mutableListOf<UserStatusChangedResponse>()
            val job = launch {
                client.subscriptions.userStatusChanged("user-1").collect { events += it }
            }
            advanceUntilIdle()

            wsAdapter.deliverConnectionAck()
            val subId = wsAdapter.getLastSubscriptionId()
            wsAdapter.deliverNext(subId, mapOf("userStatusChanged" to kUserAliceJson))
            advanceUntilIdle()

            assertEquals(1, events.size)
            assertEquals("user-1", events[0].userStatusChanged.id)
            job.cancel()
        }

        @Test
        fun event_status_isCorrect() = runTest {
            val events = mutableListOf<UserStatusChangedResponse>()
            val job = launch {
                client.subscriptions.userStatusChanged("user-1").collect { events += it }
            }
            advanceUntilIdle()

            wsAdapter.deliverConnectionAck()
            val subId = wsAdapter.getLastSubscriptionId()
            wsAdapter.deliverNext(subId, mapOf("userStatusChanged" to kUserAliceJson))
            advanceUntilIdle()

            assertEquals(UserStatus.ACTIVE, events[0].userStatusChanged.status)
            job.cancel()
        }

        @Test
        fun userId_sentInSubscribePayload() = runTest {
            val job = launch {
                client.subscriptions.userStatusChanged("user-42").collect {}
            }
            advanceUntilIdle()
            wsAdapter.deliverConnectionAck()

            val subscribeMsg = wsAdapter.sentMessagesDecoded.first { it["type"] == "subscribe" }
            @Suppress("UNCHECKED_CAST")
            val payload = subscribeMsg["payload"] as Map<String, Any?>
            @Suppress("UNCHECKED_CAST")
            val variables = payload["variables"] as Map<String, Any?>
            assertEquals("user-42", variables["userId"])
            job.cancel()
        }
    }
}
